toolchain/parse/extract.cpp

```diff
@@ -0,0 +1,334 @@
+    }
+
+    // Use a fold over the `=` operator to parse fields from right to left.
+    [[maybe_unused]] int unused;
+    static_cast<void>(((std::get<Index>(fields) =
+                            Extractable<T>::Extract(tree, it, end, trace),
+                        unused) = ... = 0));
+
+    if (!(std::get<Index>(fields).has_value() && ...)) {
+      if (trace) {
+        *trace << "Tuple: error\n";
+      }
+      return std::nullopt;
+    }
+
+    if (trace) {
+      *trace << "Tuple: success\n";
+    }
+    return std::tuple<T...>{std::move(std::get<Index>(fields).value())...};
+  }
+
+  static auto Extract(const Tree* tree, Tree::SiblingIterator& it,
+                      Tree::SiblingIterator end, ErrorBuilder* trace)
+      -> std::optional<std::tuple<T...>> {
+    return ExtractImpl(tree, it, end, trace,
+                       std::make_index_sequence<sizeof...(T)>());
+  }
+};
+
+// Extract the fields of a simple aggregate type.
+template <typename T>
+struct Extractable {
+  static_assert(std::is_aggregate_v<T>, "Unsupported child type");
+  static auto ExtractImpl(const Tree* tree, Tree::SiblingIterator& it,
+                          Tree::SiblingIterator end, ErrorBuilder* trace)
+      -> std::optional<T> {
+    if (trace) {
+      *trace << "Aggregate: begin\n";
+    }
+    // Extract the corresponding tuple type.
+    using TupleType = decltype(StructReflection::AsTuple(std::declval<T>()));
+    auto tuple = Extractable<TupleType>::Extract(tree, it, end, trace);
+    if (!tuple.has_value()) {
+      if (trace) {
+        *trace << "Aggregate: error\n";
+      }
+      return std::nullopt;
+    }
+
+    if (trace) {
+      *trace << "Aggregate: success\n";
+    }
+    // Convert the tuple to the struct type.
+    return std::apply(
+        [](auto&&... value) {
+          return T{std::forward<decltype(value)>(value)...};
+        },
+        *tuple);
+  }
+
+  static auto Extract(const Tree* tree, Tree::SiblingIterator& it,
+                      Tree::SiblingIterator end, ErrorBuilder* trace)
+      -> std::optional<T> {
+    static_assert(!HasKindMember<T>, "Missing Id suffix");
+    return ExtractImpl(tree, it, end, trace);
+  }
+};
+
+template <typename T>
+auto Tree::TryExtractNodeFromChildren(
+    llvm::iterator_range<Tree::SiblingIterator> children,
+    ErrorBuilder* trace) const -> std::optional<T> {
+  auto it = children.begin();
+  auto result = Extractable<T>::ExtractImpl(this, it, children.end(), trace);
+  if (it != children.end()) {
+    if (trace) {
+      *trace << "Error: " << node_kind(*it) << " node left unconsumed.";
+    }
+    return std::nullopt;
+  }
+  return result;
+}
+
+// Manually instantiate Tree::TryExtractNodeFromChildren
+#define CARBON_PARSE_NODE_KIND(KindName)                    \
+  template auto Tree::TryExtractNodeFromChildren<KindName>( \
+      llvm::iterator_range<Tree::SiblingIterator> children, \
+      ErrorBuilder * trace) const -> std::optional<KindName>;
```

Comment:
```suggestion
      ErrorBuilder* trace) const -> std::optional<KindName>;
```
I guess `clang-format` undoes this?

Reply:
Indeed our autoformating undoes this.